[sayomaki] iP

sayomaki opened the pull request:
# Alexer
> The future depends on what you do today - Mahatma Gandhi

A lexer that can do (**quite**) a few things. (is this a pun? maybe)
It takes care of:
- tracking your completed tasks
- watching your un-completed tasks
- ensuring you don't forget your tasks

Most importantly, it is:
- [x] *very* easy to use
- [x] simple to understand
- [ ] complicated to setup

So what are you waiting for? Download and try today! 
1. Clone the [repo](https://github.com/sayomaki/ip) or download the jar
2. Compile the code (if needed): `./gradlew shadowJar`
3. Run it and enjoy! 😄 

---

Also, here is a quick way to create multi-line text in Java! Check this out!
```java
String longLongText = """
     wee
     woo

     I am a multiline text, yipee!""";
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

sayomaki committed:
Rename Duke to Alexer and add some features

Alexer will be able to greet users and say goodbye.

sayomaki committed:
Create scanner and add prompt loop

This is to prepare for adding the echo feature, which
requires continuous reading of user input until the text
"bye" has been entered.

sayomaki committed:
Make scanner final

The scanner will be reused throughout the lifetime of the chatbot

sayomaki committed:
Add method to greet user and create an instance of chatbot

This is so that we can add more methods and
state to the chatbot later on.

sayomaki committed:
Move printing goodbye message to own function

This will make the code more modular in the long term.

sayomaki committed:
Change function ordering for chatbot actions

Ideally functions should be defined above where they are used.

sayomaki committed:
Create list for simple tasks and function to print them

This uses a Java Collection (ArrayList) to store some simple tasks with
only the name, as well as a function to print these tasks numbered.

sayomaki committed:
Add logic to create new tasks and list them

This adds some cool new features to the chatbot.

sayomaki committed:
Refactor adding task logic to its own function

This improves code abstraction and separation of concerns

sayomaki committed:
Break up input into to commands and use switch case

Using switch case will allow for more flexibility as the number
of actions increases, and separating the command out from the
input allows better parsing for future use.

sayomaki committed:
Create task class that describes a task

Our initial task consist of a task description, as well
as a boolean field to indicate whether the task is done.

There are also some helper methods to abstract commonly
used logic to make life easier later on.

sayomaki committed:
Convert primitive task list to use Task class

This offers more extensibility to tasks, such as added the task state
(done/not done) and possibly different types of tasks in future.

sayomaki committed:
Make chatbot more lively

Add more interesting prompts and responses from the chatbot

sayomaki committed:
Improve parsing of command and its arguments

This is to better prepare for command argument usage

sayomaki committed:
Add command to mark task as done

For this naive implementation all inputs are assumed to be
valid and error handling will be performed in a later part.

sayomaki committed:
Include functionality to unmark done tasks as not done

If you can mark tasks as done, why can't you unmark them?

sayomaki committed:
Return nothing as marking task is an action

sayomaki committed:
Create more specialised subclasses of Task

This will be used for future implementations

sayomaki committed:
List out the number of tasks after adding a task

sayomaki committed:
Add specialization to the Task subclasses

Deadlines now have a deadline parameter, and events have 2 new
parameters indicating the time from the event start to the event end.

sayomaki committed:
Improve code formatting

Move the newline part into a separate line.

sayomaki committed:
Add ability to create todos in the chatbot

This is similar to addTask() but instead parses
 the command arguments directly.

sayomaki committed:
Include indentation when printing out tasks

sayomaki committed:
Add ability to add deadlines to the task list

The addDeadline() function will process the input arguments to
know the deadline and description of the task, and create the
deadline task accordingly.

sayomaki committed:
Allow creation of event tasks in todo list

Similar to how addDeadline() works, we parse the arguments
by finding the index where the keywords appear and
subsection the arguments accordingly.

sayomaki committed:
Remove escaped characters in chatbot wordart

This may cause issues with tests later on.

sayomaki committed:
Fix testing code to run correct class

sayomaki committed:
Update tests with input and expected output

sayomaki committed:
Add error handling for some cases

Handle the case where the command is not found
and where the todo description is empty

sayomaki committed:
Add ability to delete tasks

This is similar to how marking tasks work in terms of parsing
which task in mentioned, and removes the task accordingly

sayomaki committed:
Update tests with more test cases

Support level 5 and level 6 features

sayomaki committed:
Create an enum for different task types

This is so we can easily differentiate task types in future.

sayomaki committed:
Make a task manager to handle all task related matters

This improves our code abstractions and allow for
easier extension such as saving and loading

sayomaki committed:
Add convenience function to print chatbot response

This will allow better abstraction of printing replies.

sayomaki committed:
Refactor chatbot to use task manager class

By using task manager, we apply more OOP principles in our code. 
Added a new method in task manager as well.

sayomaki committed:
Add method to convert a task (and subtask) to a string

This is to allow saving of tasks later on.

sayomaki committed:
Create public method to save all tasks to file

This creates and overwrite the task save file to populate
 it with all the current tasks serialised.

sayomaki committed:
Add static constructor to create task from task string

This is so that we can load the task from task strings.

sayomaki committed:
Add new method to load tasks in task manager

sayomaki committed:
Handle errors when loading tasks from disk

Inform the user if we are unable to find the save file to load,
as well as when we have successfully loaded all the saved tasks.

sayomaki committed:
Fix missing newline after each task string in save file

sayomaki committed:
Check if save folder exist otherwise create it

Here we use the folder "data" to save the tasks.

sayomaki committed:
Fix task manager only loading a single task

All the task strings in the save file should be processed

sayomaki committed:
Update test data and expected output

This is so that we can handle saved tasks correctly
and ensure consistent results

sayomaki committed:
Load tasks when chatbot starts and save tasks accordingly

sayomaki committed:
Merge branch 'branch-Level-7'

GabrielCWT commented on src/main/java/Alexer.java:
```diff
@@ -0,0 +1,224 @@
+
+    private final Scanner scanner = new Scanner(System.in);
+
+    private final TaskManager tasks = new TaskManager();
```
Maybe rename this to taskManager? I would think tasks would represent the tasks themselves e.g. an arraylist.

sayomaki replied:
Thanks for the suggestion, will take that into consideration!

GabrielCWT reviewed (commented):
Not really much complaints about code standard. Keep up the good work!

Green-Tea-123 commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,37 @@
+    public Deadline(String description, String deadline) {
+        super(description);
+        this.deadline = deadline;
+        type = TaskType.DEADLINE;
```
Perhaps you can align with the above statement and put `this.type` instead of just `type`. This applies for the other task subclasses as well.

sayomaki replied:
I was trying to follow the code guidelines for this! There is a clause on this: "Avoid unnecessary use of this with fields."

For `deadline` there is a conflicting name in the method parameter and as part of the object, so the explicit `this` keyword was necessary. However, for the `type` field there was no such ambiguity so I went ahead and omitted `this`!

Green-Tea-123 commented on src/main/java/Todo.java:
```diff
@@ -0,0 +1,34 @@
+
+    public static Todo fromTaskString(String taskString) {
+        String[] parts = taskString.split("\\|");
+        if (parts.length < 3)
+            return null;
```
Perhaps you should add open and close braces Egyptian style to comply with the coding standard.

sayomaki replied:
Noted, thanks for pointing this out!

Green-Tea-123 reviewed (commented):
Other than the comments I have made, your code complies with the coding standard for this course. Good job!

sayomaki committed:
Add date time processing to deadline

Show deadline tasks with formatting date time string.

sayomaki committed:
Include error handling for when an invalid date is loaded

Skip the loading of the task for invalid dates.

sayomaki committed:
Add brackets to if statement to conform to style guide

sayomaki committed:
Refactor task loading to skip tasks that failed to load

This is when the task is null, we just skip the test

sayomaki committed:
Update test with a proper date time in deadline task

sayomaki committed:
Merge branch 'branch-Level-8'

sayomaki committed:
Move setting up of chatbot to own method

This reduces the amount of code in the main function.

sayomaki committed:
Use the new printResponse method instead

Since we have already abstracted the printing method,
we shall use that instead

sayomaki committed:
Initialise fields explicitly in constructor and create instance

We create a reference of the instance itself (singleton) that can
be used by other classes to access the chatbot.

sayomaki committed:
Create new prompter class to handle bot responses and output

This class mainly serves to handle whatever the user sees on
their screen, i.e. the output of the chatbot and messages.

sayomaki committed:
Use prompter class to handle interaction responses

This will be the start of slowly using methods from
prompter class to handle bot output

sayomaki committed:
Fix method to get instance of chat bot

The method should be static so it is a class
method instead of an instance method.

sayomaki committed:
Add more public methods to access various parts of the chatbot

sayomaki committed:
Add a basic command handler and base command class

For the command class, we use an abstract class to ensure that it has to
be implemented by a subclass (which is the actual command).

The command handler is responsible for managing all the commands in the chatbot.

sayomaki committed:
Convert listing task operation to a command

Also registers the command with the command handler.

sayomaki committed:
Add method to print task list in prompter

sayomaki committed:
Fix list task command to print correctly formatted tasks

sayomaki committed:
Create the command handler instance and use it

Processing our first command!

sayomaki committed:
Update test output to match with new output

sayomaki committed:
Remove unused code left over from improvements

sayomaki committed:
Merge branch 'branch-OOP'

sayomaki committed:
Move classes into their respective packages

sayomaki committed:
Fix test scripts to compile all files recursively

sayomaki committed:
Merge branch 'branch-Package'

sayomaki committed:
Merge branch 'add-gradle-support'

sayomaki committed:
Update gradle build main class

sayomaki committed:
Improve formatting and remove error printing

sayomaki committed:
Add error printing when failure to load task

sayomaki committed:
Add unit testing for some tasks

These added tasks check for valid and invalid save inputs when
loading, as well as command handler to handle if command does
not exist

sayomaki committed:
Merge branch 'branch-JUnit'

sayomaki committed:
Ensure jar name is correct for output jar

sayomaki committed:
Convert event task timings to final

sayomaki committed:
Add javadocs to base package classes

For public methods missing javadoc, they will be deprecated 
and refactored later on so the javadocs 
will appear in their new locations.

sayomaki committed:
Add javadocs for the command subpackage

sayomaki committed:
Add javadoc for task subpackage

sayomaki committed:
Convert wildcard import to imports

sayomaki committed:
Add new method to print the list of filtered tasks

sayomaki committed:
Provide a way to obtain task description from task

sayomaki committed:
Create method to find a list of tasks containing keyword

sayomaki committed:
Add a new command to find tasks

This command has the name "find" and takes a single argument keyword to search for the task.

sayomaki committed:
Register the new command to the command handler

sayomaki committed:
Update CLI tests and ensure tests work

sayomaki committed:
Merge branch 'branch-Level-9'

sayomaki committed:
Merge branch 'branch-A-CodingStandard'

sayomaki committed:
Merge branch 'branch-A-JavaDoc'

sayomaki committed:
Add missing javadoc for newly added description getter

sayomaki committed:
Add javafx to gradle dependencies

sayomaki committed:
Create a new command to handle deletion of tasks

sayomaki committed:
Refactor chatbot to use delete command instead

sayomaki committed:
Create a basic graphical user interface to display

sayomaki committed:
Add a message box to show responses and interactions

sayomaki committed:
Add main launcher to show the graphical user interface

sayomaki committed:
Set the main class to the new GUI class

sayomaki committed:
Add code to setup and render the interface with sizing

sayomaki committed:
Add code to automatically scroll to bottom and move classes

Move the gui package to ui as it all belongs under user interface.

sayomaki committed:
Import images used for GUI avatar

sayomaki committed:
Add formatting for GUI message box

sayomaki committed:
Create response class to handle chatbot interactions

This class contains a wrapper around the response to be
displayed to the screen (for GUI) or the console (text)

sayomaki committed:
Create response class to handle chatbot interactions

This class contains a wrapper around the response to be
displayed to the screen (for GUI) or the console (text)

sayomaki committed:
Render a testing message box in GUI with user avatar

sayomaki committed:
Migrate chatbot response to use the new class

sayomaki committed:
Migrate chatbot response to use the new class

sayomaki committed:
Refactor the chatbot to separate out the text-based methods

sayomaki committed:
Refactor the main prompt loop used for text-based to its own method

This allows usage of the chatbot instance for the GUI mode,
and at the same time preserving the text-based functionality.

sayomaki committed:
Refactor the main prompt loop used for text-based to its own method

This allows usage of the chatbot instance for the GUI mode,
and at the same time preserving the text-based functionality.

sayomaki committed:
Add ability to flip the message box when it is created by the bot

This adds 2 static constructors to create a message box for the
respective roles in the chat bot, i.e. the user and the bot.

The bot's dialog also has the image and text order
reversed to be different than the user.

sayomaki committed:
Process user input for GUI and render on the screen

sayomaki committed:
Add new main window class for GUI with FXML

This uses FXML to define our window and render
regions instead of doing it programatically.

sayomaki committed:
Refactor the GUI application to use FXML and reduce code amount

sayomaki committed:
Fix double execution of prompt loop method

sayomaki committed:
Setup and start the chat bot for GUI mode

sayomaki committed:
Refactor commands to return a response

The response contains the output of the chat bot
that can then be displayed back to the user.

sayomaki committed:
Refactor the chatbot to separate out the text-based methods

sayomaki committed:
Fix double execution of prompt loop method

sayomaki committed:
Convert unknown command message to a response

sayomaki committed:
Print the command response after executing

sayomaki committed:
Remove unnecessary string format

sayomaki committed:
Convert the mark task feature into its own command

sayomaki committed:
Convert unmark task to its own command

sayomaki committed:
Refactor add task commands to their own commands

sayomaki committed:
Return the response on processing user input for chatbot

sayomaki committed:
Merge branch 'master' into branch-Level-10

sayomaki committed:
Remove printing of greeting and logo from start up

sayomaki committed:
Add chat bot response to GUI instead of placeholder input

sayomaki committed:
Use FXML for message box rendering

sayomaki committed:
Merge branch 'branch-Level-10'

sayomaki committed:
Use varargs for command arguments

sayomaki committed:
Improve delete task command response code format

sayomaki committed:
Remove unnecessary code and categorise variables together

sayomaki committed:
Add proper comments and logical spacing between method calls

This formats the code and improve styling for all non-command
classes.

sayomaki committed:
Add javadoc for all command classes

sayomaki committed:
Move bot logo text to prompter class so it is more organised

sayomaki committed:
Refactor strings in prompter to static constants

sayomaki committed:
Convert prompter building methods to static methods

sayomaki committed:
Use static constant values to store bot responses

This is so that we have all the response messages in one place so it
is easy to change and modify if necessary.

sayomaki committed:
Use abstraction provided by the task manager class instead

This is so that the tasks are better managed by the task manager
class, and if there are any changes to the way the tasks are created,
it can be done via the task manager

sayomaki committed:
Move responsibility of saving tasks to task manager

sayomaki committed:
Add task storage class to handle saving and loading tasks

This allows for better abstraction and handling of responsibilities
within the chatbot, and allows further extensibility with different
types of task storages that can be done by just extending this class.

sayomaki committed:
Add javadoc comments for all task manager methods

sayomaki committed:
Use the new task storage class and an option to toggle persistence

sayomaki committed:
Add some assertions to ensure correct state of code

sayomaki committed:
Create more assertions to validate states

sayomaki committed:
Reduce method size by abstracting into smaller methods

sayomaki committed:
Add continuous integration via Github Actions

This will test the textual UI interface of the chatbot automatically.

sayomaki committed:
Merge pull request #2 from sayomaki/branch-A-Assertions

Add Java assertions to Alexer

sayomaki committed:
Merge pull request #3 from sayomaki/branch-A-CodeQuality

Improve the code quality of Alexer

sayomaki committed:
Fix CI not working due to missing dos2unix

sayomaki committed:
Ensure formatted date for deadline has capital AM/PM

This is because when testing on linux systems the output date time
string is shown as uppercase, whereas on my local system (macos) 
it is shown as lowercase.

sayomaki committed:
Update textual UI tests with up-to-date output

sayomaki committed:
Allow running of CI testing on other platforms even when one fails

sayomaki committed:
Create a help command that shows

This is currently hard-coded for now, but in future it may instead
compute itself from the list of registered commands accordingly,
and be auto-updated whenever new commands are added.

sayomaki committed:
Register the help command with the command handler

sayomaki committed:
Convert help string to a text block and fix indentation for examples

sayomaki committed:
Merge branch 'branch-C-Help'

sayomaki committed:
Set the anchor for the text input box and send button to allow resizing

sayomaki committed:
Add minimum width and height for the chatbot

sayomaki committed:
Add some style to the GUI based on the guide

This is to make our chatbot look more lively and customised!

sayomaki committed:
Include input placeholder for text field box

sayomaki committed:
Change colour scheme of the GUI to make it have more personality

sayomaki committed:
Set the name of the window to the chatbot name

sayomaki committed:
Reduce display picture size and add slightly more padding

sayomaki committed:
Add padding and margin to message box and update colours

sayomaki committed:
Format the message response with additional padding to align them

sayomaki committed:
Reduce message box top and bottom padding to look more compact

sayomaki committed:
Set minimum width of GUI window to a larger value by default

sayomaki committed:
Align the help text with the start of the multi-line string

sayomaki committed:
Make response class immutable and properties public

sayomaki committed:
Allow changing the type of response and fix display of response

sayomaki committed:
Change colour of the message box containing user input

sayomaki committed:
Make message box take in a response for bot output instead

This will make it easier to parse and process the response not only
based on the string of the response, but also other properties of
the response instance.

sayomaki committed:
Change the default border width and colour for the message box

sayomaki committed:
Add custom error styling for the message box if response is an error

sayomaki committed:
Fix name of the bot image

sayomaki committed:
Add greeting to the chatbot when it is launched

sayomaki committed:
Convert unknown command response to error response type

sayomaki committed:
Remove unused method for alexer

sayomaki committed:
Add some text animations to bot responses

sayomaki committed:
Increase font size slightly and change window default size

sayomaki committed:
Merge branch 'branch-A-BetterGUI'

sayomaki committed:
Mark responses as error responses for when there is an error

sayomaki committed:
Add checking of missing event details for event tasks

sayomaki committed:
Ensure that task to be deleted is valid and exists

sayomaki committed:
Improve error handling for marking and unmarking tasks

sayomaki committed:
Add a response for when finding tasks with no tasks at all

sayomaki committed:
Merge branch 'branch-A-MoreErrorHandling'

sayomaki committed:
Create another test for registering new commands

sayomaki committed:
Add product introduction to user guide

sayomaki committed:
Add screenshot to user guide

sayomaki committed:
Complete the user guide with all the features
